spec/models/category_spec.rb

```diff
@@ -1550,4 +1550,17 @@
       )
     end
   end
+
+  describe ".limited_categories_matching" do
+    before_all { SiteSetting.max_category_nesting = 3 }
+
+    fab!(:foo) { Fabricate(:category, name: "foo") }
+    fab!(:bar) { Fabricate(:category, name: "bar", parent_category: foo) }
+    fab!(:baz) { Fabricate(:category, name: "baz", parent_category: bar) }
+
+    it "produces results in depth-first pre-order" do
+      SiteSetting.max_category_nesting = 3
+      expect(Category.limited_categories_matching(nil, "baz").pluck(:name)).to eq(%w[foo bar baz])
```

Comment:
I'm sure you're thinking about this, but we need a couple more tests here:

- what happens when user searches for "bar"? I would expect the parent to be included (foo), but not the child? 
- what happens when there is no term included? I'd expect all categories to load, right?